### PR TITLE
OCPBUGS-99659: reconcile NetworkPolicies for KubeVirt HCP router mode

### PR DESCRIPTION
**Pull request by RamLavi**
## What this PR does / why we need it:

KubeVirt centralized clusters using KAS **Route + explicit `route.hostname`** (HCP router / full network isolation) deploy an HCP `private-router` behind a LoadBalancer. The hypershift-operator reconciled platform NetworkPolicies for AWS/Azure/GCP but not for KubeVirt, leaving two gaps:

1. **`private-router` NetworkPolicy** was never created — LoadBalancer traffic to router pods (`app=private-router`) could be denied by namespace default-deny ingress rules.
2. **`virt-launcher` NetworkPolicy** egress allowed kube-apiserver, oauth, and ignition-server-proxy peers but not the HCP router — worker VMs reaching Ignition via the MetalLB-backed router hit the virt-launcher IPBlock `Except` list.

This PR:
- Reconciles the `private-router` NetworkPolicy for centralized KubeVirt when `UseDedicatedDNSForKASByHC` is true (and deletes it when the HCP router is not in use).
- Adds `app=private-router` as a virt-launcher egress peer under the same condition.

External-infra KubeVirt and default LoadBalancer-based KubeVirt clusters are unchanged.

## Which issue(s) this PR fixes:

Fixes OCPBUGS-99659

## Manual validation (KubeVirt centralized + KAS Route hostname)

Lab: hitchhiker-02, mgmt OCP 4.20, CNV + MetalLB, HostedCluster `ralavi-hg` in HCP router mode (`api-ralavi-hg.qinqon.corp`). Workers not required for netpol proof.

HO image tested: `quay.io/ramlavi/hypershift-operator:OCPBUGS-99659-ed04981`

### BEFORE (stock `quay.io/hypershift/hypershift-operator:latest`)

```
=== HostedCluster status ===
NAME        VERSION   KUBECONFIG                   PROGRESS   AVAILABLE   PROGRESSING   MESSAGE
ralavi-hg             ralavi-hg-admin-kubeconfig   Partial    True        False         The hosted control plane is available
=== NodePool / machines ===
NAME        CLUSTER     DESIRED NODES   CURRENT NODES   AUTOSCALING   AUTOREPAIR   VERSION   UPDATINGVERSION   UPDATINGCONFIG   MESSAGE
ralavi-hg   ralavi-hg   1                               False         False                  True              True             
=== HCP router mode indicators ===
KAS service: {"route":{"hostname":"api-ralavi-hg.qinqon.corp"},"type":"Route"}
Note: on public mgmt clusters the HCP router LB service is named router (pods still use app=private-router)
NAME     TYPE           CLUSTER-IP       EXTERNAL-IP       PORT(S)         AGE
router   LoadBalancer   172.30.127.208   192.168.122.242   443:31585/TCP   153m
Error from server (NotFound): services "private-router" not found
=== NetworkPolicy: private-router (expect NotFound BEFORE fix) ===
Error from server (NotFound): networkpolicies.networking.k8s.io "private-router" not found
=== NetworkPolicy: virt-launcher egress to private-router (expect NO match BEFORE fix) ===
virt-launcher egress private-router peer: NO
=== virt-launcher pods ===
```

### AFTER (fixed HO, ~2 min reconcile)

```
=== HostedCluster status ===
NAME        VERSION   KUBECONFIG                   PROGRESS   AVAILABLE   PROGRESSING   MESSAGE
ralavi-hg             ralavi-hg-admin-kubeconfig   Partial    True        False         The hosted control plane is available
=== NodePool / machines ===
NAME        CLUSTER     DESIRED NODES   CURRENT NODES   AUTOSCALING   AUTOREPAIR   VERSION   UPDATINGVERSION   UPDATINGCONFIG   MESSAGE
ralavi-hg   ralavi-hg   1                               False         False                  True              True             
=== HCP router mode indicators ===
KAS service: {"route":{"hostname":"api-ralavi-hg.qinqon.corp"},"type":"Route"}
Note: on public mgmt clusters the HCP router LB service is named router (pods still use app=private-router)
NAME     TYPE           CLUSTER-IP       EXTERNAL-IP       PORT(S)         AGE
router   LoadBalancer   172.30.127.208   192.168.122.242   443:31585/TCP   3h11m
Error from server (NotFound): services "private-router" not found
=== NetworkPolicy: private-router (expect NotFound BEFORE fix) ===
NAME             POD-SELECTOR         AGE
private-router   app=private-router   110s
=== NetworkPolicy: virt-launcher egress to private-router (expect NO match BEFORE fix) ===
virt-launcher egress private-router peer: YES
=== virt-launcher pods ===
```

Note: `Service/private-router` NotFound on both runs is **expected** on a public management cluster — only the public LB service is named `router`; pods still use `app=private-router`. The bug is the missing **NetworkPolicy**, not the Service.

## Special notes for your reviewer:

- Two focused commits: ingress NP first, virt-launcher egress second.
- Unit tests added in `network_policies_test.go` for both behaviors.

## Checklist:
- [x] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.
- [ ] This change includes docs.
- [x] This change includes unit tests.

## Summary by CodeRabbit

- **Bug Fixes**
  - Corrected network policy handling for centralized KubeVirt clusters using dedicated HCP routing.
  - Added separate IPv4 and IPv6 egress rules so network exceptions are applied to the correct address family.
  - Ensured private-router access is permitted for virt-launcher workloads when required.
  - Removed stale private-router policy behavior when the API server uses a load balancer instead of dedicated routing.
  - Improved handling of management-cluster network ranges and API server access exceptions in dual-stack environments.

### CODE REVIEW

File: hypershift-operator/controllers/hostedcluster/network_policies.go [hunk 1]
```diff
@@ -178,6 +178,20 @@ func (r *HostedClusterReconciler) reconcilePlatformNetworkPolicies(ctx context.C
 			}); err != nil {
 				return fmt.Errorf("failed to reconcile virt launcher policy: %w", err)
 			}
+			if kubevirtCentralizedUsesHCPRouter(hcluster) {
+				policy := networkpolicy.PrivateRouterNetworkPolicy(controlPlaneNamespaceName)
+				ingressOnly := version.Major == 4 && version.Minor < 14
+				if _, err := createOrUpdate(ctx, r.Client, policy, func() error {
+					return reconcilePrivateRouterNetworkPolicy(policy, hcluster, kasBlock, r.ManagementClusterCapabilities.Has(capabilities.CapabilityDNS), managementClusterNetwork, ingressOnly)
```
Comment (qinqon) [inline hypershift-operator/controllers/hostedcluster/network_policies.go:185]: This enables an IPv4-only policy builder for KubeVirt. On dual-stack management clusters, it puts IPv6 pod CIDRs into the `Except` list of `0.0.0.0/0`, which Kubernetes rejects, causing HostedCluster reconciliation to fail repeatedly. 

Please split exclusions into matching IPv4/IPv6 IPBlocks and add a dual-stack test.

**Review by qinqon — CHANGES_REQUESTED** [review]

File: hypershift-operator/controllers/hostedcluster/network_policies.go (hunk 1 above)
Comment (RamLavi) [reply hypershift-operator/controllers/hostedcluster/network_policies.go:185]: good catch. fixed

**Comment by RamLavi** [comment]
[Change](https://github.com/openshift/hypershift/compare/5c17e4c418d4e4fa54a73122982afb7663cb8c66..e06b304c9848995e3a84b359d1dd627373b64cd9):  split exclusions into matching IPv4/IPv6 IPBlocks and add a dual-stack test.

**Comment by RamLavi** [comment]
[Change](https://github.com/openshift/hypershift/compare/e06b304c9848995e3a84b359d1dd627373b64cd9..c491068f2dd218296036bcc90c456a51fce7d8e7): Rebase (no conflicts)

File: hypershift-operator/controllers/hostedcluster/network_policies.go [hunk 2]
```diff
@@ -178,6 +178,20 @@ func (r *HostedClusterReconciler) reconcilePlatformNetworkPolicies(ctx context.C
...
+				}); err != nil {
+					return fmt.Errorf("failed to reconcile private router network policy: %w", err)
+				}
+			} else {
+				policy := networkpolicy.PrivateRouterNetworkPolicy(controlPlaneNamespaceName)
+				if _, err := k8sutil.DeleteIfNeeded(ctx, r.Client, policy); err != nil {
```
Comment (jparrill) [inline hypershift-operator/controllers/hostedcluster/network_policies.go]: nit: `policy` is created identically in both branches. You could pull it above the conditional to avoid the duplication:

```go
policy := networkpolicy.PrivateRouterNetworkPolicy(controlPlaneNamespaceName)
if kubevirtCentralizedUsesHCPRouter(hcluster) {
    ingressOnly := version.Major == 4 && version.Minor < 14
    if _, err := createOrUpdate(ctx, r.Client, policy, func() error {
        ...
```

File: hypershift-operator/controllers/hostedcluster/network_policies.go [hunk 3]
```diff
@@ -340,27 +354,34 @@ func reconcilePrivateRouterNetworkPolicy(policy *networkingv1.NetworkPolicy, _ *
 		return nil
 	}
 
-	clusterNetworks := make([]string, 0)
+	blockedIPv4Networks := make([]string, 0, len(kasBlockExceptions))
```
Comment (jparrill) [inline hypershift-operator/controllers/hostedcluster/network_policies.go:356]: This dual-stack refactor is a good fix — IPv6 CIDRs in the `Except` list of a `0.0.0.0/0` IPBlock were silently ignored by kube-proxy/OVN.

However, `reconcilePrivateRouterNetworkPolicy` is shared by all platforms (AWS/Azure/GCP call it at line 168), so this change adds an explicit `::/0` IPv6 egress peer to the private-router NetworkPolicy for **all platforms**, not just KubeVirt.

The behavioral change is correct and additive (IPv6 egress blocking that previously did nothing now actually works), but it would be worth mentioning this cross-platform scope in the PR description so reviewers are aware of the broader blast radius.

File: hypershift-operator/controllers/hostedcluster/network_policies_test.go [hunk 4]
```diff
@@ -952,6 +1018,166 @@ func TestReconcilePlatformNetworkPolicies(t *testing.T) {
 	}
 }
 
+func TestKubevirtCentralizedUsesHCPRouter(t *testing.T) {
```
Comment (jparrill) [inline hypershift-operator/controllers/hostedcluster/network_policies_test.go:1032]: suggestion: consider adding a test case where `Kubevirt` field is nil on a KubeVirt-platform HC.

Right now all cases set `Kubevirt: &hyperv1.KubevirtPlatformSpec{}`. If `Kubevirt` is nil, the guard `Kubevirt != nil && Credentials != nil` evaluates to false and the function falls through to `UseDedicatedDNSForKASByHC` — which could return true if Route+hostname is set. An explicit case would document whether that's intended behavior or just an unreachable state.

```go
{
    name: "When platform is KubeVirt with nil Kubevirt spec and Route+hostname, it should return true",
    hcluster: &hyperv1.HostedCluster{
        Spec: hyperv1.HostedClusterSpec{
            Platform: hyperv1.PlatformSpec{Type: hyperv1.KubevirtPlatform},
            Services: []hyperv1.ServicePublishingStrategyMapping{...Route+hostname...},
        },
    },
    want: true, // or false — whichever is the intended contract
}
```

File: hypershift-operator/controllers/hostedcluster/network_policies_test.go [hunk 5]
```diff
@@ -854,6 +890,35 @@ func TestReconcilePlatformNetworkPolicies(t *testing.T) {
...
+			services: []hyperv1.ServicePublishingStrategyMapping{
+				{
+					Service: hyperv1.APIServer,
+					ServicePublishingStrategy: hyperv1.ServicePublishingStrategy{
+						Type:  hyperv1.Route,
+						Route: &hyperv1.RoutePublishingStrategy{Hostname: "api.example.com"},
+					},
+				},
+			},
+			expectPrivateRouter: true,
```
Comment (jparrill) [inline hypershift-operator/controllers/hostedcluster/network_policies_test.go:907]: suggestion: this case verifies the NP isn't *created* when `expectPrivateRouter: false`, but it doesn't test the `DeleteIfNeeded` cleanup path. If you pre-seed a `private-router` NetworkPolicy into the fake client before reconciling, you can assert it gets deleted — that would cover the else branch in the production code.

File: hypershift-operator/controllers/hostedcluster/network_policies.go (hunk 2 above)
Comment (RamLavi) [reply hypershift-operator/controllers/hostedcluster/network_policies.go]: you're right, will fix.

File: hypershift-operator/controllers/hostedcluster/network_policies.go (hunk 3 above)
Comment (RamLavi) [reply hypershift-operator/controllers/hostedcluster/network_policies.go:356]: I agree. Added a "Dual-stack scope" section to the PR desc.

File: hypershift-operator/controllers/hostedcluster/network_policies_test.go (hunk 4 above)
Comment (RamLavi) [reply hypershift-operator/controllers/hostedcluster/network_policies_test.go:1032]: Added it, with a small concern - that documenting an unsupported behavior - makes it supported a little :) 
But on second thought - better to know how this thing will behave in this case

File: hypershift-operator/controllers/hostedcluster/network_policies_test.go (hunk 5 above)
Comment (RamLavi) [reply hypershift-operator/controllers/hostedcluster/network_policies_test.go:907]: yeah I agree, added a pre-seeding mechanism for our case, tell me what you think

**Comment by RamLavi** [comment]
[Change](https://github.com/openshift/hypershift/compare/c491068f2dd218296036bcc90c456a51fce7d8e7..cbb5fe8a674518be7c74b7ab33cefa3e6c37408d): Address @jparrill review

**Comment by RamLavi** [comment]
[Change](https://github.com/openshift/hypershift/compare/cbb5fe8a674518be7c74b7ab33cefa3e6c37408d..3494f7f981f2a5818484950fe929677b83eee158): fix valid CR [issue](https://github.com/openshift/hypershift/pull/9619#pullrequestreview-5289916650)

/coderabbit resolve

**Comment by RamLavi** [comment]
@jparrill I understand that there's a HCP QE capacity issue
Is my [Manual validation] section in PR desc satisfactory in order for me to self verify?

**Comment by jparrill** [comment]
> Is my [Manual validation] section in PR desc satisfactory in order for me to self verify?

@RamLavi We usually ask for an external verification, someone not related with the bug. But your validation would be ok if you can cover the whole behavior of the PR.

One concern: in both runs the HC stays Partial with 0 worker nodes, so there's no functional traffic flowing through the NetworkPolicy. The bug is that LB traffic to router pods gets denied — but without at least one worker up, we can only confirm the NP is created with the right spec, not that it actually unblocks the traffic.

Could you re-run with at least 1 worker node up and show that ignition/oauth are reachable through the router? That would close the loop on functional verification.
